virtualfish/virtual.fish

```diff
@@ -129,13 +169,92 @@ function __vf_new --description "Create a new virtualenv"
     end
 
     emit virtualenv_will_create
-    set envname $argv[-1]
-    set -e argv[-1]
-    if set -q VIRTUALFISH_DEFAULT_PYTHON
-        set argv "--python" $VIRTUALFISH_DEFAULT_PYTHON $argv
+    argparse -n "vf new" -x q,v,d --ignore-unknown "h/help" "q/quiet" "v/verbose" "d/debug" "p/python=" -- $argv
+
+    if set -q _flag_help
+        echo "Purpose: Creates a new virtual environment"
+        echo "Usage: "(set_color green)"vf new "(set_color -di)"[-p <python-version>] [<virtualenv-flags>]"(set_color normal)(set_color green)" <virtualenv-name>"(set_color normal)
+        echo
+        echo "Examples:"
+        echo
+        echo (set_color green)"vf new -p /usr/local/bin/python3 yourproject"(set_color normal)
+        echo (set_color green)"vf new -p python3.8 --system-site-packages yourproject"(set_color normal)
+        echo
+        echo "To see available "(set_color blue)"Virtualenv"(set_color normal)" option flags, run: "(set_color green)"virtualenv --help"(set_color normal)
+        return 0
+    end
+
+    # Unpack Virtualenv args: first flags that need values, then Boolean flags
+    # When done, $argv should only contain the new virtual environment name.
+    set -l flags_with_args "--app-data" "--discovery" "--creator" "--seeder" "--activators" "--extra-search-dir" "--pip" "--setuptools" "--wheel" "--prompt"
+    for arg in $argv
+        # If arg starts with a hyphen…
+        if test (string sub -l 1 -- $arg) = "-"
+            # Retrieve its index number
+            set -l idx (contains -i -- $arg $argv)
+            # If this option requires a value that we expect to come after it…
+            if contains -- $arg $flags_with_args
+                # Get that value's index number
+                set -l next_arg_idx (math $idx + 1)
+                # Move both the option flag and its value to a separate list
+                set virtualenv_args $virtualenv_args $arg $argv[$next_arg_idx]
+                set -e argv[$next_arg_idx]
```

Comment:
You'll still loop over this with the current for-loop. $argv was already expanded when that was executed the first time.

So if an option-argument starts with a "-"....

Reply:
I can see how there could be an issue if someone passed a hyphen-prefixed value to an option flag, but as noted below, I couldn't quite figure out how to modify the for-loop. This feels like a big enough edge case that I’m not sure it’s worth spending too much time on it, but if it were easy to address, of course, I’m all for it! 😅